android/modules/database/src/java/ti/modules/titanium/database/TiDatabaseProxy.java

```diff
@@ -18,145 +19,244 @@
 import android.content.Context;
 import android.database.Cursor;
 import android.database.DatabaseUtils;
-import android.database.SQLException;
 import android.database.sqlite.SQLiteDatabase;
+import android.os.Looper;
+
+import java.util.ArrayList;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 @Kroll.proxy(parentModule = DatabaseModule.class)
 public class TiDatabaseProxy extends KrollProxy
 {
 	private static final String TAG = "TiDB";
 
+	private static Thread thread;
+	private static BlockingQueue<Runnable> queue = new LinkedBlockingQueue<>();
+	private static AtomicBoolean executingQueue = new AtomicBoolean(false);
+
 	protected SQLiteDatabase db;
 	protected String name;
-	boolean statementLogging, readOnly;
 
+	/**
+	 * Create a TiDatabaseProxy for SQLiteDatabase instance.
+	 * @param name Database name.
+	 * @param db SQLiteDatabase instance.
+	 */
 	public TiDatabaseProxy(String name, SQLiteDatabase db)
 	{
 		super();
+
+		if (this.thread == null) {
+
+			// Query execution thread.
+			this.thread = new Thread(new Runnable() {
```

Comment:
when do you clean up this thread? I assume once a DB is closed, this should be interrupted and stop running.

Reply:
The idea was to have a single background thread for all TiDatabase instances, to conserve order (Multiple TiDatabase proxies can refer to the same database). But on second thought, it's not necessary anymore. So I've reverted to a thread for each TiDatabase instance.